app/src/main/java/me/hufman/androidautoidrive/phoneui/viewmodels/CarDrivingStatsModel.kt

```diff
@@ -1,252 +1,454 @@
+	val speedDisplayed = carInfo.cdsData.liveData[CDS.DRIVING.SPEEDDISPLAYED].map {
+		it.tryAsJsonPrimitive("speedDisplayed")?.tryAsDouble
+	}.format("%.0f").addUnit(unitsAverageSpeedLabel)
+	val tempInterior = carInfo.cdsData.liveData[CDS.SENSORS.TEMPERATUREINTERIOR].map {
+		it.tryAsJsonPrimitive("temperatureInterior")?.tryAsDouble
+	}.format("%.1f").addUnit(unitsTemperatureLabel)
+	val tempExterior = carInfo.cdsData.liveData[CDS.SENSORS.TEMPERATUREEXTERIOR].map {
+		it.tryAsJsonPrimitive("temperatureExterior")?.tryAsDouble
+	}.format("%.1f").addUnit(unitsTemperatureLabel)
+
+	/*
+		0 - Initialisierung -> initialization
+		1 - Tractionmodus -> Traction mode
+		2 - Komfortmodus -> Comfort mode
+		3 - Basismodus -> Basic mode
+		4 - Sportmodus -> Sports mode
+		5 - Sportmodusplus -> Sport+
+		6 - Racemodus -> Race
+		7 - Ecopro ->EcoPro
+		8 - Ecoproplus -> EcoPro+
+		9 - Komfortmoduserweitert -> Comfort mode extended
+		xx - Adaptive Mode
+		15 - Unknown
+
+		Some cars have "+" modes (sport+, comfort+) and some other the same modes are called "Individual".
+	 */
+	//val drivingMode = carInfo.cdsData.liveData[CDS.DRIVING.MODE].map {
+	val drivingMode = carInfo.cachedCdsData.liveData[CDS.DRIVING.MODE].map {
+		val a = it.tryAsJsonPrimitive("mode")?.tryAsInt
+		if (a != null){
+			if(a == 2)
+				"Comfort"
+			else if (a == 9)
+				"Comfort+"
+			else if (a == 4)
+				"Sport"
+			else if (a== 5)
+				"Sport+"
+			else if(a==6)
+				"Race"
+			else if(a==7)
+				"EcoPro"
+			else if(a==8)
+				"EcoPro+"
+			else
+				"-"
+		}
+		else
+			"-"
+		//"DEBUG: $a"
+
+	}
+	/*
+		2 - handbrake
+		0 - Bremsensymbol_AUS -> Brake symbol off
+		1 - Bremsensymbol_Gelb -> yellow
+		2 - Bremsensymbol_Rot -> red
+		3 - Bremsemsymbol_Gruen -> green
+		4 - PARK_Gelb
+		8 - PARK_Rot
+		12 - Park_Gruen
+		16 - AutoP_Gelb
+		32 - AutoP_Rot
+		48 - AutoP_Gruen
+		127 - Unknown
+	 */
+
+	val parkingBrake = carInfo.cachedCdsData.liveData[CDS.DRIVING.PARKINGBRAKE].map {
+		val pB = it.tryAsJsonPrimitive("parkingBrake")?.tryAsInt
+		if (pB != null && pB == 2) {
+			"HandBrake ON"
+		}
+		else if (pB != null && (pB == 8 || pB == 32)) {
+			"AutoPark Brake ON"
+		}
+		else
+			""
+
+		//"$pB"
+	}
+
+	/*
+		status: 0 - closed
+		status: 1 - partially open  (not tilted)
+		status: 2 - fully open
+		tilt: 1->12 -> tilted (tilted degree)
+		open: 1-50 -> how far is open
+
+	 */
+	val sunRoof =carInfo.cachedCdsData.liveData[CDS.CONTROLS.SUNROOF].map {
+		val status = it.tryAsJsonObject("sunroof")?.tryAsJsonPrimitive("status")?.tryAsInt
+		val openPosition = it.tryAsJsonObject("sunroof")?.tryAsJsonPrimitive("openPosition")?.tryAsInt
+		val tiltPosition = it.tryAsJsonObject("sunroof")?.tryAsJsonPrimitive("tiltPosition")?.tryAsInt
+		var sunRoofString = ""
+		if (status == 0 && tiltPosition != null && tiltPosition>0)
+		{
+			sunRoofString = "Open, tilted"
+		}
+		if (status != null && status == 1 && openPosition != null && openPosition >0) {

```

Comment:
You do need to check for !=null for the greater-than comparison, but the direct equality checks in these if-blocks can be simplified a little bit:
```
if (status == 0 && tiltPosition != null && tiltPosition>0)
if (status == 1 && openPosition != null && openPosition >0)
if (status == 2)
if (status == 0 && tiltPosition == 0 && openPosition == 0)
```

Alternatively, you could change the json parsing to do `?.tryAsInt ?: 0` and then all the greater-than checks can skip the null checks.

Reply:
done